api/gin/webServer.go

```diff
@@ -0,0 +1,239 @@
+package gin
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	logger "github.com/ElrondNetwork/elrond-go-logger"
+	"github.com/ElrondNetwork/elrond-go/api/address"
+	"github.com/ElrondNetwork/elrond-go/api/block"
+	"github.com/ElrondNetwork/elrond-go/api/gin/disabled"
+	"github.com/ElrondNetwork/elrond-go/api/hardfork"
+	"github.com/ElrondNetwork/elrond-go/api/middleware"
+	"github.com/ElrondNetwork/elrond-go/api/network"
+	"github.com/ElrondNetwork/elrond-go/api/node"
+	"github.com/ElrondNetwork/elrond-go/api/shared"
+	"github.com/ElrondNetwork/elrond-go/api/transaction"
+	valStats "github.com/ElrondNetwork/elrond-go/api/validator"
+	"github.com/ElrondNetwork/elrond-go/api/vmValues"
+	"github.com/ElrondNetwork/elrond-go/api/wrapper"
+	"github.com/ElrondNetwork/elrond-go/config"
+	"github.com/ElrondNetwork/elrond-go/core/check"
+	"github.com/ElrondNetwork/elrond-go/facade"
+	"github.com/ElrondNetwork/elrond-go/marshal"
+	"github.com/gin-contrib/cors"
+	"github.com/gin-contrib/pprof"
+	"github.com/gin-gonic/gin"
+)
+
+var log = logger.GetOrCreate("api/factory")
+
+// ArgsNewWebServer holds the arguments needed to create a new instance of webServer
+type ArgsNewWebServer struct {
+	Facade          shared.ApiFacadeHandler
+	ApiConfig       config.ApiRoutesConfig
+	AntiFloodConfig config.WebServerAntifloodConfig
+}
+
+type webServer struct {
+	facade          shared.ApiFacadeHandler
+	mutFacadeSwitch sync.RWMutex
+	apiConfig       config.ApiRoutesConfig
+	antiFloodConfig config.WebServerAntifloodConfig
+	httpServer      shared.HttpServerCloser
+	ctx             context.Context
+	cancelFunc      func()
```

Comment:
cancelFunc doesn't seem to be called by anyone.

I would change a bit the implementation in the following way:

- have webServer to also implement Close
- webServer Close will call cancelFunc() & httpServer.Close
- createInitialHttpServer doesn't need to return both wrapper and wrapped servers, it can just return the wrapper (webServer)
- createApiFacade doesn't need to have arguments both wrapper and closable/wrapped HTTPServer, as you can implement a getter for the httpserver on the wrapper, and it doesn't need to return any closableHttpServer, as you can access it through the getter.
- to the waitForSignal give the wrapper instead of the httpserver as argument, as the wrapper Close function needs to be called here instead (which will in turn call also httpserver close - closeAllComponents will do the wrapper Close)

Reply:
done. now the wrapper will handle also the closing of the http server.